components/request_otr/browser/request_otr_rule.cc

```diff
@@ -0,0 +1,121 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/request_otr/browser/request_otr_rule.h"
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "base/json/json_reader.h"
+#include "base/types/expected.h"
+#include "components/prefs/pref_service.h"
+#include "extensions/common/url_pattern.h"
+#include "net/base/url_util.h"
+#include "url/gurl.h"
+#include "url/origin.h"
+#include "url/url_constants.h"
+
+namespace {
+
+// request-otr.json keys
+const char kInclude[] = "include";
+const char kExclude[] = "exclude";
+
+}  // namespace
+
+namespace request_otr {
+
+RequestOTRRule::RequestOTRRule() {}
+
+RequestOTRRule::~RequestOTRRule() = default;
+
+// static
+bool RequestOTRRule::GetURLPatternSetFromValue(
+    const base::Value* value,
+    extensions::URLPatternSet* result) {
+  if (!value->is_list()) {
+    return false;
+  }
+  std::string error;
+  bool valid = result->Populate(
+      value->GetList(), URLPattern::SCHEME_HTTP | URLPattern::SCHEME_HTTPS,
+      false, &error);
+  if (!valid) {
+    DVLOG(1) << error;
+  }
+  return valid;
+}
+
+// static
+void RequestOTRRule::RegisterJSONConverter(
+    base::JSONValueConverter<RequestOTRRule>* converter) {
+  converter->RegisterCustomValueField<extensions::URLPatternSet>(
+      kInclude, &RequestOTRRule::include_pattern_set_,
+      GetURLPatternSetFromValue);
+  converter->RegisterCustomValueField<extensions::URLPatternSet>(
+      kExclude, &RequestOTRRule::exclude_pattern_set_,
+      GetURLPatternSetFromValue);
+}
+
+// static
+const std::string RequestOTRRule::GetETLDForRequestOTR(
```

Comment:
I think it's a bit confusing to have this method because it makes it seem like the ETLD for RequestOTR is somehow different than normal, but it's just calling `GetDomainAndRegistry` which is called directly basically everywhere else. If we wanted to do something like this I think we'd have a more generic helper and use it everywhere

Reply:
Some services call GetDomainAndRegistry with private registries included and some with private registries excluded. We've run into this issue before with debounce and decided that each service should have an explicit function like this to make it clear that not all services use the same parameters and are not interchangeable.

Reply:
can you add some comments to that effect and explain the reason for exclude in this case? I wasn't even aware that we were using different values for that and it's easy to miss.

Reply:
tracked in https://github.com/brave/brave-browser/issues/30306